src/Microsoft.ML.Mkl.Components/VectorWhitening.cs

```diff
@@ -617,6 +620,57 @@ protected override Delegate MakeGetter(DataViewRow input, int iinfo, Func<int, b
                     };
                 return del;
             }
+            public bool CanSaveOnnx(OnnxContext ctx) => true;
+
+            public void SaveAsOnnx(OnnxContext ctx)
+            {
+                Host.CheckValue(ctx, nameof(ctx));
+                int numColumns = _parent.ColumnPairs.Length; //why are the multiple columns
+                for (int iinfo = 0; iinfo < numColumns; ++iinfo)
+                {
+                    string inputColumnName = _parent.ColumnPairs[iinfo].inputColumnName;
+                    if (!ctx.ContainsColumn(inputColumnName))
+                        continue;
+
+                    string outputColumnName = _parent.ColumnPairs[iinfo].outputColumnName;
+                    string srcVariableName = ctx.GetVariableName(inputColumnName);
+                    string dstVariableName = ctx.AddIntermediateVariable(_srcTypes[iinfo], outputColumnName, true);
+                    SaveAsOnnxCore(ctx, iinfo, srcVariableName, dstVariableName);
+                }
+
+            }
+            private void SaveAsOnnxCore(OnnxContext ctx, int iinfo, string srcVariableName, string dstVariableName)
+            {
+                var model = _parent._models[iinfo];
+                int dimension = _srcTypes[iinfo].GetValueCount();
+                Host.Assert(model.Length == dimension * dimension);
+
+                var parameters = _parent._columns[iinfo];
+                Host.Assert(parameters.Kind == WhiteningKind.PrincipalComponentAnalysis || parameters.Kind == WhiteningKind.ZeroPhaseComponentAnalysis);
+
+                int rank = (parameters.Kind == WhiteningKind.PrincipalComponentAnalysis && parameters.Rank > 0) ? parameters.Rank : dimension;
+                long[] modelDimension = { rank, dimension };
+
+                if (rank != dimension)
+                {
+                    float[] principalComponents = new float[rank * dimension];
+                    for (int i = 0; i < parameters.Rank; i++)
```

Comment:
>for (int i = 0; i < parameters.Rank; i++) [](start = 20, length = 41)

If I understand correctly, this loop copies the first `rank*dimension` elements from `model` to `principalComponents` in the same order. Do we really need to copy?
I noticed that `ctx.AddInitializer` takes an `IEnumerable`, perhaps it would be enough to call it with `model.Take(rank*dimension)`?

Reply:
I think this problem with Take is that when the rank is higher than the dimension, ML.NET simply treats the extra eigenvectors as rows of zeros. For ONNX, the model tensor has to be resized. I can introduce a check for both ML.NET and ONNX to check that the rank < dimension?